### PR TITLE
Zcnsc latest

### PR DESCRIPTION
**Pull request by Dmdv**
This adds a new smart contract for converting between ERC-20 ethereum tokens and ZCN. It adds six new endpoints defined in sc.go. Details in 0Chain Token Bridge document.

image

### CODE REVIEW

File: code/go/0chain.net/smartcontract/storagesc/allocation_test.go [hunk 1]
```diff
@@ -1390,8 +1390,7 @@ func Test_finalize_allocation(t *testing.T) {
 	var tx = newTransaction(b1.id, ssc.ID, 0, tp)
 	balances.setTransaction(t, tx)
 	var resp string
-	resp, err = ssc.commitBlobberConnection(tx, mustEncode(t, &cc),
-		balances)
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/storagesc/allocation_test.go:1394]: As there is no meaninful change and its nothing to do with zcnsc, its probably  best to remove allocation_test.go from the PR.

File: code/go/0chain.net/chaincore/state/mint.go [hunk 2]
```diff
@@ -9,14 +9,13 @@ import (
...
 
-//Transfer - a data structure to hold state transfer from one client to another
 type Mint struct {
 	Minter     datastore.Key `json:"minter"`
 	ToClientID datastore.Key `json:"to"`
 	Amount     Balance       `json:"amount"`
 }
 
-//NewTransfer - create a new transfer
+// NewMint NewTransfer - create a new transfer
```
Comment (Sriep) [inline code/go/0chain.net/chaincore/state/mint.go:18]: New mint does not create a transfer, it mints tokens.

File: code/go/0chain.net/smartcontract/zcnsc/models.go [hunk 3]
```diff
@@ -0,0 +1,497 @@
...
+	if err == nil {
+		return gn
+	}
+
+	gn.MinMintAmount = state.Balance(config.SmartContractConfig.GetInt("smart_contracts.zcn.min_mint_amount"))
+	gn.PercentAuthorizers = config.SmartContractConfig.GetFloat64("smart_contracts.zcn.percent_authorizers")
+	gn.MinAuthorizers = config.SmartContractConfig.GetInt64("smart_contracts.zcn.min_authorizers")
+	gn.MinBurnAmount = config.SmartContractConfig.GetInt64("smart_contracts.zcn.min_burn_amount")
+	gn.MinStakeAmount = config.SmartContractConfig.GetInt64("smart_contracts.zcn.min_stake_amount")
+	gn.BurnAddress = config.SmartContractConfig.GetString("smart_contracts.zcn.burn_address")
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/zcnsc/models.go:84]: I am confused. The global node is retrieved from the blockchain, then all the fields are overwritten. Why the blockchain access?

File: code/go/0chain.net/smartcontract/zcnsc/sc.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	return zcnCopy
+}
+
+// InitSC ...
+func (zcn *ZCNSmartContract) InitSC() {}
+
+// SetSC ...
+func (zcn *ZCNSmartContract) setSC(sc *smartcontractinterface.SmartContract, bcContext smartcontractinterface.BCContextI) {
+	zcn.SmartContract = sc
+	zcn.SmartContract.RestHandlers["/globalPerodicLimit"] = zcn.globalPerodicLimit
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/zcnsc/sc.go:40]: Typo `Periodic`. What is periodic about the cofig values returned?

File: code/go/0chain.net/smartcontract/zcnsc/models.go [hunk 5]
```diff
@@ -0,0 +1,497 @@
...
+		return
+	}
+	an.NodeMap[node.ID] = node
+	return
+}
+
+func getAuthorizerNodes(balances cstate.StateContextI) (*authorizerNodes, error) {
+	an := &authorizerNodes{}
+	av, err := balances.GetTrieNode(allAuthorizerKey)
+	if err != nil {
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/zcnsc/models.go:400]: I think best style would be somthing like
```go
if err != nil {
  if err != unit.ErrValueNotPresent {
        return nil, err
  } else {
        an.NodeMap = make(map[string]*authorizerNode)
        return an, nil  
  }
}

```

File: docker.local/bin/unit_test_build.sh [hunk 6]
```diff
@@ -1,9 +1,26 @@
 #!/bin/bash
 set -e
 
+cmd="build"
+dockerfile="docker.local/build.unit_test/Dockerfile"
+platform=""
+
+for arg in "$@"
```
Comment (Sriep) [inline docker.local/bin/unit_test_build.sh:8]: What are these blocks added to the `docker.build.*.sh` files all about?

File: docker.local/build.unit_test/Dockerfile.m1 [hunk 7]
```diff
@@ -0,0 +1,24 @@
+FROM --platform=linux/amd64 zchain_build_base
```
Comment (Sriep) [inline docker.local/build.unit_test/Dockerfile.m1:1]: Why is this needed?

**Review by Sriep — CHANGES_REQUESTED** [review]
This adds a new smart contract for converting between ERC-20 ethereum tokens and ZCN. It adds six new endpoints defined in sc.go. Details in 0Chain Token Bridge document.

This PR also makes a lot of changes to docker.local. Not sure why?

Unit tests. Better to use `mocks.StateContextI`. It allows input and output values to and from StateContextI to be monitored.

File: code/go/0chain.net/smartcontract/zcnsc/mint_test.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
+package zcnsc
+
```
[the remark this answers is not in the document]
Comment (Dmdv) [reply code/go/0chain.net/smartcontract/zcnsc/mint_test.go:2]: I added mint tests

File: code/go/0chain.net/smartcontract/zcnsc/burn.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+	payload.TxnID = t.Hash
+
+	// check burn amount
+	if t.Value < gn.MinBurnAmount {
+		err = common.NewError("failed to burn", fmt.Sprintf("amount requested(%v) is lower than min amount for burn (%v)", t.Value, gn.MinBurnAmount))
+		return
+	}
+
+	payload.Amount = t.Value
```
[the remark this answers is not in the document]
Comment (Dmdv) [reply code/go/0chain.net/smartcontract/zcnsc/burn.go]: Why it is ignored? We use it as a response. Could you explain your point 
@Sriep

File: code/go/0chain.net/smartcontract/storagesc/allocation_test.go (hunk 1 above)
Comment (Dmdv) [reply code/go/0chain.net/smartcontract/storagesc/allocation_test.go:1394]: Yes, just strange formatting, I changed it accidentally

File: code/go/0chain.net/chaincore/state/mint.go (hunk 2 above)
Comment (Dmdv) [reply code/go/0chain.net/chaincore/state/mint.go:18]: Agree, not my comment. I fixed the IDE warning, I shouldn't change the code not related to PR

File: code/go/0chain.net/smartcontract/zcnsc/test_factory_helpers.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+		BurnAddress:        "0",
+		MinAuthorizers:     1,
+	}
+}
+
+func createBurnPayload() *burnPayload {
+	return &burnPayload{
+		TxnID:           txHash,
+		Nonce:           1,
+		Amount:          100,
```
[the remark this answers is not in the document]
Comment (Dmdv) [reply code/go/0chain.net/smartcontract/zcnsc/test_factory_helpers.go:102]: The above conversation - we need Amount as it is used in burn ticket

File: code/go/0chain.net/smartcontract/zcnsc/models.go (hunk 3 above)
Comment (Dmdv) [reply code/go/0chain.net/smartcontract/zcnsc/models.go:62]: @Sriep This is repeated after many other places, for example, in miners, the globalNode after fetching is being updated according to configuration

File: docker.local/bin/unit_test_build.sh (hunk 6 above)
Comment (Dmdv) [reply docker.local/bin/unit_test_build.sh:8]: Like everything in master branch, this is m1 support

File: docker.local/build.unit_test/Dockerfile.m1 (hunk 7 above)
Comment (Dmdv) [reply docker.local/build.unit_test/Dockerfile.m1:1]: Removed as not needed any longer

File: code/go/0chain.net/smartcontract/zcnsc/models.go (hunk 5 above)
Comment (Dmdv) [reply code/go/0chain.net/smartcontract/zcnsc/models.go:408]: Fixed

File: code/go/0chain.net/smartcontract/zcnsc/sc.go (hunk 4 above)
Comment (Dmdv) [reply code/go/0chain.net/smartcontract/zcnsc/sc.go]: Makes sense. Removed as useless

File: code/go/0chain.net/smartcontract/zcnsc/test_context_helpers.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+		&state.Deserializer{},
+		txn,
+		nil,
+		nil,
+		nil,
+		nil,
+	)
+}
+
+func CreateMockStateContext(clientId string) cstate.StateContextI {
```
[the remark this answers is not in the document]
Comment (Dmdv) [reply code/go/0chain.net/smartcontract/zcnsc/test_context_helpers.go:116]: In progress too replace it with mocks

File: code/go/0chain.net/smartcontract/zcnsc/models.go (hunk 3 above)
Comment (Sriep) [reply code/go/0chain.net/smartcontract/zcnsc/models.go:62]: Sorry, I think its ok, leave as is.

File: code/go/0chain.net/smartcontract/zcnsc/burn.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (Sriep) [reply code/go/0chain.net/smartcontract/zcnsc/burn.go]:
```go
// inputData - is a burnPayload
func (zcn *ZCNSmartContract) burn(trans *transaction.Transaction, inputData []byte, balances cstate.StateContextI) (resp string, err error) {
        ...
	payload := &burnPayload{}
	err = payload.Decode(inputData)
...
	payload.TxnID = trans.Hash
...
       payload.Amount = trans.Value
...
      resp = string(payload.Encode())
```
As far as I can tell payload is only used in `burn`. 
1. You put the input into a `burnPaload` object
2. You overwrite the payload.TxnID field.
3. You overwrite the payload.Amount field
4. Do some stuff setting the TxnID and Amount fields
5. Return a string version of payload.

So the payload.TxnID and payload.Amount from the input are not used. The fields are only used to create a return string.
Anyone creating input for `burn` is going to hit a learning curve as they will initially expect the Amount and TxnId field to mean something to the `burm` method. So I was suggesting removing it from the input object. Just use them to create the return string.

File: code/go/0chain.net/smartcontract/zcnsc/test_factory_helpers.go [hunk 12]
```diff
@@ -0,0 +1,173 @@
+package zcnsc
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/zcnsc/test_factory_helpers.go:1]: If this file is only used for testing, use the `_test` suffix so it is not part of the production build.

**Review by Sriep — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/smartcontract/zcnsc/test_factory_helpers.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Sriep) [reply code/go/0chain.net/smartcontract/zcnsc/test_factory_helpers.go:102]: My point is that it is not used in `burn`. Can the value you set `burnPayload.Amount` or `burnPayload.TaxID` have any effect on anything? I may be wrong so could you give like to where it is used - I mean the value, not the field.

File: code/go/0chain.net/smartcontract/zcnsc/test_factory_helpers.go (hunk 12 above)
Comment (Dmdv) [reply code/go/0chain.net/smartcontract/zcnsc/test_factory_helpers.go:1]: Moved to test module

File: code/go/0chain.net/smartcontract/zcnsc/test_context_helpers.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (Dmdv) [reply code/go/0chain.net/smartcontract/zcnsc/test_context_helpers.go]: This is fixed

File: code/go/0chain.net/chaincore/state/mint.go (hunk 2 above)
Comment (Sriep) [inline code/go/0chain.net/chaincore/state/mint.go]: New mint does not create a new transfer

File: code/go/0chain.net/smartcontract/zcnsc/README.MD [hunk 13]
```diff
@@ -0,0 +1,4 @@
+![Diagram](https://github.com/Dmdv/0chain/blob/zcnsc_latest/code/go/0chain.net/smartcontract/zcnsc/docs/ZCNSC.png?raw=true)
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/zcnsc/README.MD:1]: Nice pictures, but maybe add some text to introduce them?

File: docker.local/bin/build.miners.sh [hunk 14]
```diff
@@ -9,6 +9,8 @@ DOCKER_DIR="$ROOT/docker.local/build.miner"
 DOCKER_FILE="$DOCKER_DIR/Dockerfile"
 DOCKERCOMPOSE="$DOCKER_DIR/docker-compose.yml"
 
+APP_ROOT="$ROOT/code/go/0chain.net"
```
Comment (Sriep) [inline docker.local/bin/build.miners.sh]: Does not seem to be used.

**Comment by Dmdv** [comment]
I will close because I already have https://github.com/0chain/0chain/pull/536 merged into staging
